---
title: "103-make-reference seqs"
author: "Mac Campbell"
date: "June 2, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```


## I need to create seqs

n flanks and then replace position with [Major/Minor].     

Wrote extractFlanksandRecode.pl, should work on outputs/101/snps-75-samples-recode-filtered.MAF.txt

```{sh}
srun -p high -t 1:00:00 --nodes=1  extractFlanksandRecode.pl $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa outputs/101/snps-75-samples-recode-filtered.MAF.txt >> outputs/103/snps-75-samples-recode-filtered.fasta
```

It generally works, but I need to get to singleton sites, that, is no nearby snps.

```{r}
library(tidyverse)
```

```{r}
data<-read_tsv("outputs/101/snps-75-samples-recode-filtered.MAF.txt", col_names = c("Chrom","Site","Major","Minor","MAF")) %>%
  group_by(Chrom)
```

```{r}
data %>% mutate(Lag=Site-lag(Site)) %>% mutate(Lead=Site-lead(Site))
```


```{r}
filt<-data %>% mutate(Lag=Site-lag(Site)) %>% mutate(Lead=abs(Site-lead(Site))) %>% 
  filter(Lag > 1000) %>%
  filter(Lead > 1000)

filt
```

This gets us 1212 sites.

```{r}
write_tsv(filt, "meta/snps-75-samples-recode-filtered.MAF.singletons.txt", col_names = FALSE)
```

```{sh}
srun -p high -t 1:00:00 --nodes=1  extractFlanksandRecode.pl $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa $HOME/dsm-genotyping/meta/snps-75-samples-recode-filtered.MAF.singletons.txt > outputs/103/snps-75-samples-recode-filtered.fasta
```

Takes forever.   I wonder if I can do this chrom by chrom by using samtools faidx in the script to avoid loading to memory.

```{sh, eval=FALSE}
./103-extract.pl meta/test.singletons.txt
```

Removing [M/m] and blasting
`lg01-site-3438302-Major-G-Minor-T-MAF=0.138462
TCCTCCTCTCTTCCTCTCCTCCCTTCCTCTCCTCCTCTCCTCCTCTCTTCCTCTCCCCCTCTCCCGCTCTCTCCCTCTCCTCCTCCCTTCCTCTCCTCCTCTCCTCGTCGGCCAGGTGCCACAGGTGTGGAGGACCGTCTGCAGGAGAGCGCCCCTGACACCATCGCTGCTCTCCGAGAGGCGGGGATCCAGGTGTGGGTGCTTACAGGCGACAAGGCCGAGACGGCCGTCAACATCGCATACGCCTGCAGGCTGCTGGATGATGGGGACCTGGTGTTCAACATGAGAACCAGCAGCAAGG
lg01-site-3545233-Major-G-Minor-T-MAF=0.113678
TATCCCTGCAGGGTGGATCCCCTCCCTGTGAGATAACACTGGAGGGATTCAGAAGTGTGTGTGTGGCTGTGTGTGTGTGTGTGTAAATATATCTCAGCGGAGAGGGAAGGCAGCGCTGAGACAGAAGTTCTAGAAAGTACAACTGAAGCTGCCATTAATGTGTGTGACTGAGATGAACAGGTCTCTCTCTCTCTCACTCACACTCACACTCACACACAAACCCACCTTCACACTTCACACACATACAGGGTAGTGTAGGTAGGTATGGTTGTGTACTGTTGTGGTGTTCCATACAAGTTTG`

`(base) maccamp\@farm:~/genomes/hypomesus-20210204$ blastn -query test.fasta -db lg01.fasta -outfmt 6 | sort -k 3 -n
lg01-site-3438302-Major-G-Minor-T-MAF=0.138462	lg01	100.000	301	0	0	1	301	3438152	3438452	7.60e-159	556
lg01-site-3545233-Major-G-Minor-T-MAF=0.113678	lg01	100.000	301	0	0	1	301	3545083	3545383	7.60e-159	556
`
Looks ok.

```{sh, eval=FALSE}
srun -p high -t 1:00:00 --nodes=1 ./103-extract.pl meta/snps-75-samples-recode-filtered.MAF.singletons.txt  > outputs/103/snps-75-samples-recode-filtered.fasta
```

~/dsm-genotyping/outputs/103$ grep ">" snps-75-samples-recode-filtered.fasta | wc -l      
1212

Produces 1212 seqs as expected. I also think we should do soemthing about high copy number seqs by blasting against the genome.

