---
title: "200-hybrids"
author: "Mac Campbell"
date: "April 19, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message = FALSE)
```

```{r}
library(tidyverse)
```

## Other species
Finding some markers for osmerids. From Benjamin et al. I obtained the rad contigs for the assay.

Where are the raw sequences?
`/home/ajfinger/BMAG043_Wakasagi`
Set up:    
`(base) maccamp@farm:~/dsm-genotyping/data$ ls | grep WAKS | grep R1 | perl -pe 's/\.fastq//g' > waks.R1.txt`     
`(base) maccamp@farm:~/dsm-genotyping/data$ ls | grep WAKS | grep R2 | perl -pe 's/\.fastq//g' > waks.R2.txt`    
`(base) maccamp@farm:~/dsm-genotyping/data$ paste waks.R1.txt waks.R2.txt > waks.txt`     
`(base) maccamp@farm:~/dsm-genotyping/data$ ../100.1-do-align.sh waks.txt $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa`

In the meantime, let's see if we can identify some contigs from hybrids:

## Contigs
###AssayName Locus DSM WKS LFS

From Benjamin et al. (2018), created assay-seqs.fasta

Checking it out:
`(base) maccamp@farm:~/genomes/hypomesus-20210204$ makeblastdb -in Hyp_tra_F_20210204.fa -dbtype 'nucl'`

Making a file without [/d]   
`(base) maccamp@farm:~/dsm-genotyping/outputs/200$ cat assay-seqs.fasta  | perl -pe 's/\[//g' | perl -pe 's/\/\w\]//g' > seqs.fasta`   
Something like this:

`(base) maccamp@farm:~/dsm-genotyping/outputs/200$ blastn -query seqs.fasta -evalue .001 -db ~/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa -outfmt 6 > blast.out`

What do these look like?     
(base) maccamp@farm:~/dsm-genotyping/outputs/200$ cut -f 1 blast.out | sort | uniq -c
      1 DSM_01   
      2 DSM_02   
      1 DSM_03   
      1 DSM_04  
      1 DSM_05   
      1 DSM_06   
      1 DSM_07     
      1 DSM_08   
      1 DSM_09   
      1 DSM_10    
   4945 LFS_01   
     18 LFS_03    
   5546 LFS_04    
      1 LFS_05    
      3 LFS_06   
      1 LFS_07    
      1 WKS_01  
      1 WKS_02     
      1 WKS_03  
      1 WKS_04  
      1 WKS_05  
      2 WKS_06    
      1 WKS_07     
 
 Yikes!!!! DFAM has no hits, but there are a lot of hits to the reference genome....
 
 (base) maccamp@farm:~/dsm-genotyping/outputs/200$ grep LFS_01 blast.out  | sort -n -k 4 | tail     
LFS_01	lg06	89.143	175	13	6	397	569	4606702	4606872	1.36e-53	213    
LFS_01	lg07	90.857	175	9	5	394	565	9352927	9352757	4.87e-58	228    
LFS_01	lg13	91.429	175	10	4	395	565	6352161	6351988	2.91e-60	235    
LFS_01	lg21	86.286	175	8	5	395	569	5393679	5393837	1.79e-42	176    
LFS_01	scaffold_144	90.857	175	13	3	392	565	219321	219149	3.77e-59	231     
LFS_01	scaffold_184	89.714	175	12	3	396	569	96899	97068	2.93e-55	219    
LFS_01	scaffold_89	85.311	177	11	4	408	569	501553	501377	3.00e-40	169    
LFS_01	lg19	85.955	178	8	12	406	572	8929241	8929070	6.44e-42	174    
LFS_01	lg01	85.377	424	43	12	1	416	8608388	8608800	2.08e-116	422   
LFS_01	lg01	91.608	429	23	3	1	416	8514514	8514086	3.25e-164	580   

Most hits are short.

(base) maccamp@farm:~/dsm-genotyping/outputs/200$ grep LFS_04 blast.out  | sort -n -k 4 | tail     
LFS_04	lg22	87.168	226	22	5	150	373	3057252	3057472	7.87e-65	250    
LFS_04	scaffold_146	85.841	226	23	6	142	363	265646	265866	2.85e-59	231    
LFS_04	scaffold_27	85.841	226	24	6	158	382	1288207	1288425	7.92e-60	233    
LFS_04	lg01	82.969	229	30	6	158	384	1574364	1574585	2.89e-49	198    
LFS_04	lg11	85.217	230	24	8	138	363	311965	311742	3.69e-58	228    
LFS_04	scaffold_105	79.130	230	30	8	164	389	413093	412878	1.37e-32	143   
LFS_04	lg06	83.190	232	21	6	148	366	12589375	12589601	1.04e-48	196   
LFS_04	lg02	79.574	235	39	5	158	389	10115823	10115595	1.36e-37	159   
LFS_04	lg06	79.661	236	32	15	159	390	44647	44424	1.76e-36	156   
LFS_04	scaffold_359	79.325	237	32	16	159	390	74624	74400	8.21e-35	150    

A few hits are nearly the same size..... Probably should chuck it.     

(base) maccamp@farm:~/dsm-genotyping/outputs/200$ grep LFS_03 blast.out  | sort -n -k 4 | tail    
LFS_03	scaffold_60	90.858	711	41	8	4	691	198366	197657	0.0	931    
LFS_03	scaffold_60	90.858	711	41	8	4	691	20730	20021	0.0	931    
LFS_03	scaffold_60	90.858	711	41	8	4	691	242178	241469	0.0	931       
LFS_03	scaffold_60	90.858	711	41	8	4	691	273039	272330	0.0	931     
LFS_03	scaffold_60	90.858	711	41	8	4	691	71238	70529	0.0	931   
LFS_03	scaffold_60	90.858	711	41	8	4	691	80003	79294	0.0	931   
LFS_03	scaffold_60	90.858	711	41	8	4	691	88737	88028	0.0	931   
LFS_03	scaffold_60	90.858	711	41	8	4	691	98990	98281	0.0	931   
LFS_03	lg08	91.457	714	37	8	1	691	14069342	14068630	0.0	959    
LFS_03	scaffold_334	91.176	714	39	8	1	691	63913	63201	0.0	948   

Huh.  May be alright.    

Let's take a look at the tabular:    
(base) maccamp@farm:~/dsm-genotyping/outputs/200$ grep -v LFS_01 blast.out | grep -v LFS_04 | grep -v LFS_03
DSM_01	lg16	94.818	521	17	7	1	515	6424833	6424317	0.0	804    
DSM_02	lg06	93.531	371	24	0	2	372	4900430	4900060	7.85e-156	553    
DSM_02	lg09	89.552	134	8	4	452	585	13043591	13043464	4.30e-39	165      
DSM_03	lg02	91.449	573	27	7	22	593	11822070	11822621	0.0	767    
DSM_04	lg10	90.188	744	33	7	1	712	9034650	9033915	0.0	933     
DSM_05	lg16	94.626	521	18	7	1	515	6424833	6424317	0.0	798     
DSM_06	lg24	90.528	549	26	5	1	544	3275491	3274964	0.0	702    
DSM_07	scaffold_107	88.608	316	26	7	1	315	698034	697728	9.70e-103	375      
DSM_08	lg22	95.804	572	16	4	1	572	7369559	7368996	0.0	917      
DSM_09	lg01	97.114	693	16	1	1	689	7927198	7927890	0.0	1166    
DSM_10	lg07	91.682	529	24	8	1	524	3392415	3391902	0.0	715     
WKS_01	lg03	94.310	703	30	2	1	696	6044534	6043835	0.0	1068    
WKS_02	lg10	94.751	381	16	1	1	381	3412174	3412550	3.32e-167	590     
WKS_03	lg11	96.358	604	22	0	1	604	14582368	14582971	0.0	994     
WKS_04	lg07	96.438	758	25	2	1	756	8427685	8426928	0.0	1249     
WKS_05	lg01	92.000	325	4	3	1	303	7163931	7164255	3.64e-121	436     
WKS_06	lg23	91.257	366	11	3	301	645	2113706	2114071	1.31e-133	479     
WKS_06	lg23	96.341	246	9	0	1	246	2113437	2113682	2.26e-111	405        
WKS_07	lg08	92.487	386	22	2	53	437	728701	729080	8.45e-154	545      
LFS_05	lg17	95.479	376	14	3	1	375	8223117	8223490	1.95e-169	597     
LFS_06	lg12	88.964	444	28	8	1	439	8726408	8725981	1.61e-148	529     
LFS_06	lg12	89.085	284	21	6	420	699	8725936	8725659	6.27e-93	344     
LFS_06	lg12	90.071	141	2	3	666	805	8725576	8725447	3.13e-41	172    
LFS_07	lg04	96.835	316	10	0	1	316	3989846	3989531	6.00e-149	529     
 


##Overlap with called SNPS.
Putting together lengths of queries for for filtering:
```{sh, eval=FALSE}
cat seqs.fasta | awk '$0 ~ ">" {if (NR > 1) {print c;} c=0;printf substr($0,2,100) "\t"; } $0 !~ ">" {c+=length($0);} END { print c; }' > seq-lens.tsv
```

```{r}
lens<-read_tsv("outputs/200/seq-lens.tsv", col_names = FALSE) %>% separate(X1, sep=" ", into=c("query","RADcontig","geno1","geno2","geno3")) %>%
            select(query,X2) %>% rename(Length=X2)
```


```{r}
df<-read_tsv("outputs/200/blast.out", col_names = c("query", "target", "%id", "alignment-length", "diffs", "gaps", "query-start",
                                                      "query-end", "target-start", "target-end", "e-value", "bitscore")) %>%
  left_join(lens)
df %>% group_by(query, Length) %>% summarize(Count=n()) %>% arrange(-Count)
```

 Ehhhwww.....
 
 Drop out short hits, must be > 0.9 length of query       
 
```{r}
df %>% filter(`alignment-length`> 0.9*Length) %>% group_by(query) %>% summarize(Count=n()) %>% arrange(-Count)
```
Also eehhww... 
Let's be cautious about the specificity of LFS_01, LFS_03, LFS_04, and get some .5*query alignments

```{r}
ddf <- df %>% filter(`alignment-length`> .5*Length) %>% filter(!query %in% c("LFS_01","LFS_03","LFS_04")) %>%
  mutate(Start=ifelse(`target-start` < `target-end`, `target-start`, `target-end`)) %>%
          mutate(End=ifelse(`target-start` < `target-end`, `target-end`, `target-start`)) %>%
          mutate(orient=End-Start)

ddf %>% group_by(query) %>% summarize(Count=n()) %>% arrange(-Count)

write_tsv(ddf, "outputs/400/hybridization-snps-blastn.tsv")

```

How do these ddf snps relate to our SNPs of interest.

```{r}

mafs<-read_tsv("outputs/101/doPlink-90-MAF.txt", col_names=c("Chromosome","Position","Minor","Major","MAF"))  %>%
  separate(MAF, into=c("Type","Frequency"), sep="=")

dddf<-ddf %>% inner_join(mafs, by=c("target"="Chromosome")) %>% filter(Position > Start) %>% filter(Position < End)
dddf %>% select(query, Start, End, Position, Minor, Major, Type, Frequency)
write_tsv(dddf, "outputs/400/hybrid-matches.txt")
```

```{r}
mafs2<-read_tsv("outputs/300/samples-90-nothwemaf.txt", col_names=c("Chromosome","Position","Minor","Major","MAF"))  %>%
  separate(MAF, into=c("Type","Frequency"), sep="=")

dddf2<-ddf %>% inner_join(mafs2, by=c("target"="Chromosome")) %>% filter(Position > Start) %>% filter(Position < End)
dddf2 %>% select(query, Start, End, Position, Minor, Major, Type, Frequency)
write_tsv(dddf2, "outputs/400/hybrid-matches-nomaf-nohwe.tsv")
```

This is great, as we don't want polymorphic sites.

