---
title: "300-parentage"
author: "Mac Campbell"
date: "5/4/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
```

## Parentage
Lew et al. 2015 has some SNPs. 
`(base) Macs-MacBook-Pro-2:300 mac$ cat lew-et-al-seqs.fasta| perl -pe 's/\[//g' | perl -pe 's/\/\w\]//g' > lew.fasta`

Now to blast against the DSM genome.     
`blastn -query lew.fasta -evalue .001 -db ~/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa -outfmt 6 > blast.out`    

These look great!!!    
Htr-GVL-A002185 lg01    100.000 138     0       0       1       138     10040190        10040327        4.39e-67        255        
Htr-GVL-A000163 lg02    100.000 138     0       0       1       138     11995053        11994916        4.39e-67        255        
Htr-GVL-A000314 lg03    100.000 138     0       0       1       138     4916411 4916548 4.39e-67        255        
Htr-GVL-A000225 lg04    100.000 138     0       0       1       138     5353012 5352875 4.39e-67        255        
Htr-GVL-A002099 lg05    99.275  138     1       0       1       138     5380318 5380181 2.04e-65        250        
Htr-GVL-A000492 lg06    99.275  138     1       0       1       138     11726616        11726753        2.04e-65        250        
Htr-GVL-A000612 lg09    100.000 138     0       0       1       138     9907223 9907086 4.39e-67        255        
Htr-GVL-A001701 lg11    100.000 138     0       0       1       138     4134056 4134193 4.39e-67        255        
Htr-GVL-A000505 lg12    100.000 138     0       0       1       138     6611724 6611587 4.39e-67        255        
Htr-GVL-A002206 lg13    100.000 138     0       0       1       138     5302157 5302294 4.39e-67        255        
Htr-GVL-A001430 lg14    100.000 138     0       0       1       138     16464411        16464274        4.39e-67        255        
Htr-GVL-A001499 lg15    98.551  138     2       0       1       138     2271636 2271499 9.51e-64        244        
Htr-GVL-A000714 lg17    99.275  138     1       0       1       138     1892822 1892959 2.04e-65        250        
Htr-GVL-A001852 lg18    99.275  138     1       0       1       138     6597240 6597103 2.04e-65        250        
Htr-GVL-A000500 lg19    99.275  138     1       0       1       138     6907154 6907291 2.04e-65        250        
Htr-GVL-A000078 lg20    100.000 138     0       0       1       138     3143798 3143661 4.39e-67        255        
Htr-GVL-A002161 lg21    100.000 138     0       0       1       138     6308702 6308565 4.39e-67        255        
Htr-GVL-A000651 lg22    100.000 138     0       0       1       138     3326448 3326585 4.39e-67        255        
Htr-GVL-A001702 lg23    99.275  138     1       0       1       138     1607135 1607272 2.04e-65        250        
Htr-GVL-A002071 lg24    99.275  138     1       0       1       138     3903943 3903806 2.04e-65        250        
Htr-GVL-A000850 lg26    99.275  138     1       0       1       138     2158667 2158804 2.04e-65        250        
      
These are great, but they aren't being used at the moment. I need to find the 74 actual snps being used now..

Turns out these are all the SNPs from lew et al:
`(base) Macs-MBP-2:300 mac$ cut -f 4,5,10 -d ',' lew\ assay.csv | perl -pe 's/^H/\>H/g' | perl -pe 's/(\d),([A|G|C|T])/$1\n$2/g' > lew-all-seqs.fasta`     
`(base) Macs-MBP-2:300 mac$ grep ">" lew-all-seqs.fasta | wc -l`     
      75     
      
Checking against genome....
`cat lew-all-seqs.fasta | perl -pe 's/\[//g' | perl -pe 's/\/\w\]//g' >lew-all-seqs-seq-only.fasta`    
`blastn -query lew-all-seqs-seq-only.fasta -evalue .001 -db ~/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa -outfmt 6 > blast.out`

For the most part these look good, a few are absent from the assembly.   

Here is one that shows duplications:       
query target %id alignment-length diffs gaps query-start query-end target-start target-end e-value bitscore      
Htr-GVL-A001607,Htr-GVL-059 lg10 99.275 138 1 0 1 138 6705721 6705584 2.04e-65 250      

Htr-GVL-A001607,Htr-GVL-059 lg26 97.080 137 4 0 2 138 2779390 2779254 7.40e-60 231      

Htr-GVL-A001607,Htr-GVL-059 scaffold_62 94.074 135 7 1 135 959253 959119 1.61e-51 204      

This appears to be an orf:    
>lcl|ORF1    
MTQFHLINQRSLEDIIQHLKTSSCR    

Hard to align to proteins or tblastx against refseq rna for a clear match, matches mostly unknown/hypothesized proteins.      

It is not in the dfam database, btws....    


## Coordinate spaces of these SNPS
Want to check out how these relate to our genotype calls.

```{r}
blast<-read_tsv("outputs/300/blast.out", col_names = c("query", "target", "%id", "alignment-length", "diffs", "gaps", "query-start",
                                                      "query-end", "target-start", "target-end", "e-value", "bitscore"))

write_tsv(blast, "outputs/400/parentage-snps-blastn.tsv")
```

Again, we have only 65 placed on the DSM genome. 

```{r}
#mafs<-read_tsv("outputs/101/doPlink-90-MAF.txt", col_names=c("Chromosome","Position","Minor","Major","MAF"))  %>%
 # separate(MAF, into=c("Type","Frequency"), sep="=")

mafs<-read_tsv("outputs/102/top90-high-norm-filtered.MAF.txt", col_names=c("Chromosome","Position","Minor","Major","MAF"))  %>%
  separate(MAF, into=c("Type","Frequency"), sep="=")


```


We want target = Chromosome, bounded by target-start/target-end, these may be flipped.

```{r}
blast2<-blast %>% mutate(Start=ifelse(`target-start` < `target-end`, `target-start`, `target-end`)) %>%
          mutate(End=ifelse(`target-start` < `target-end`, `target-end`, `target-start`)) %>%
          mutate(orient=End-Start)
```

Should be able to use multiple conditions, perhaps by joining then filtering.

```{r}
matches<-blast2 %>% inner_join(mafs, by=c("target"="Chromosome")) %>% filter(Position > Start) %>% filter(Position < End)
matches %>% select(query, Start, End, Position, Minor, Major, Type, Frequency)
write_csv(matches, "outputs/400/parentage-matches.csv")
```

What about not filtering of MAF/HWE?


```{r}
mafs2<-read_tsv("outputs/102/top90-high-norm-light-filtered.MAF.txt", col_names=c("Chromosome","Position","Minor","Major","MAF"))  %>%
  separate(MAF, into=c("Type","Frequency"), sep="=")
```


```{r}
matches2<-blast2 %>% inner_join(mafs2, by=c("target"="Chromosome")) %>% filter(Position > Start) %>% filter(Position < End)
matches2 %>% select(query, Start, End, Position, Minor, Major, Type, Frequency)
write_csv(matches2, "outputs/400/parentage-matches-light-filter-bcftools.csv")
```

Looks like 